ICE on debian 32 bit using f64

I tried the simplest 32bit target, debian-i386 and it builds, but make
check-rust fails. It seems to have trouble processing f64
defintions/values.

### Code

see: gccrs/gcc/testsuite/rust.test/compile/methods1.rs

### Meta

- latest master.

### Error output

```
/home/mark/gccrs/gcc/testsuite/rust.test/compile/methods1.rs:38:9: warning: unused name 'sum'
   38 |     let sum = rect.sum_x();
      |         ^

Analyzing compilation unit
/home/mark/gccrs/gcc/testsuite/rust.test/compile/methods1.rs: In function ‘Rectangle_sum_x’:
/home/mark/gccrs/gcc/testsuite/rust.test/compile/methods1.rs:26:5: error: type mismatch in binary expression
   26 |     fn sum_x(self) -> f64 {
      |     ^
f64

<float:80>

<float:80>

D.227 = _2 + _4;
/home/mark/gccrs/gcc/testsuite/rust.test/compile/methods1.rs:26:5: internal compiler error: ‘verify_gimple’ failed
0x8b114e4 verify_gimple_in_seq(gimple*)
        ../../gccrs/gcc/tree-cfg.c:5157
0x881a9e3 gimplify_body(tree_node*, bool)
        ../../gccrs/gcc/gimplify.c:15401
0x881abb2 gimplify_function_tree(tree_node*)
        ../../gccrs/gcc/gimplify.c:15472
0x8675c88 cgraph_node::analyze()
        ../../gccrs/gcc/cgraphunit.c:670
0x8678a58 analyze_functions
        ../../gccrs/gcc/cgraphunit.c:1236
0x8679641 symbol_table::finalize_compilation_unit()
        ../../gccrs/gcc/cgraphunit.c:2514
Please submit a full bug report,
with preprocessed source if appropriate.
Please include the complete backtrace with any bug report.
See <https://gcc.gnu.org/bugs/> for instructions.

```

<!--
Include a backtrace if available.
-->
<details><summary>Backtrace</summary>
  <p>
  
  ```
  /home/mark/gccrs/gcc/testsuite/rust.test/compile/methods1.rs:26:5: internal compiler error: ‘verify_gimple’ failed
0x8b114e4 verify_gimple_in_seq(gimple*)
        ../../gccrs/gcc/tree-cfg.c:5157
0x881a9e3 gimplify_body(tree_node*, bool)
        ../../gccrs/gcc/gimplify.c:15401
0x881abb2 gimplify_function_tree(tree_node*)
        ../../gccrs/gcc/gimplify.c:15472
0x8675c88 cgraph_node::analyze()
        ../../gccrs/gcc/cgraphunit.c:670
0x8678a58 analyze_functions
        ../../gccrs/gcc/cgraphunit.c:1236
0x8679641 symbol_table::finalize_compilation_unit()
        ../../gccrs/gcc/cgraphunit.c:2514
  ```
  
  </p>
</details>

--- a/gcc/rust/rust-gcc.cc
+++ b/gcc/rust/rust-gcc.cc
@@ -1940,14 +1940,16 @@ Gcc_backend::negation_expression (NegationOperator op, Bexpression *expr,
   /* For negation operators, the resulting type should be the same as its
      operand. */
   auto tree_type = TREE_TYPE (expr_tree);
+  auto original_type = tree_type;
   auto tree_code = operator_to_tree_code (op);
 
   /* For floating point operations we may need to extend the precision of type.
      For example, a 64-bit machine may not support operations on float32. */
   bool floating_point = is_floating_point (expr_tree);
+  auto extended_type = NULL_TREE;
   if (floating_point)
     {
-      auto extended_type = excess_precision_type (tree_type);
+      extended_type = excess_precision_type (tree_type);
       if (extended_type != NULL_TREE)
 	{
 	  expr_tree = convert (extended_type, expr_tree);
@@ -1958,6 +1960,8 @@ Gcc_backend::negation_expression (NegationOperator op, Bexpression *expr,
   /* Construct a new tree and build an expression from it. */
   auto new_tree = fold_build1_loc (location.gcc_location (), tree_code,
 				   tree_type, expr_tree);
+  if (floating_point && extended_type != NULL_TREE)
+    new_tree = convert (original_type, expr_tree);
   return this->make_expression (new_tree);
 }
 
@@ -1982,13 +1986,15 @@ Gcc_backend::arithmetic_or_logical_expression (ArithmeticOrLogicalOperator op,
   /* For arithmetic or logical operators, the resulting type should be the same
      as the lhs operand. */
   auto tree_type = TREE_TYPE (left_tree);
+  auto original_type = tree_type;
   auto tree_code = operator_to_tree_code (op, floating_point);
 
   /* For floating point operations we may need to extend the precision of type.
      For example, a 64-bit machine may not support operations on float32. */
+  auto extended_type = NULL_TREE;
   if (floating_point)
     {
-      auto extended_type = excess_precision_type (tree_type);
+      extended_type = excess_precision_type (tree_type);
       if (extended_type != NULL_TREE)
 	{
 	  left_tree = convert (extended_type, left_tree);
@@ -2000,6 +2006,8 @@ Gcc_backend::arithmetic_or_logical_expression (ArithmeticOrLogicalOperator op,
   /* Construct a new tree and build an expression from it. */
   auto new_tree = fold_build2_loc (location.gcc_location (), tree_code,
 				   tree_type, left_tree, right_tree);
+  if (floating_point && extended_type != NULL_TREE)
+    new_tree = convert (original_type, new_tree);
   return this->make_expression (new_tree);
 }
 

